### PR TITLE
Brought alpha-blended mesh rendering in line with opaque mesh rendering.

### PR DESCRIPTION
**Pull request by immortius**
Fix for invisible translucent mesh bug as reported by @skaldarnar 

There are still some rendering issues related the interaction of translucent mesh and depth blur.
zesphere

### CODE REVIEW

**Comment by emanuele3d** [comment]
Hmmm.. obviously here the level of blur is multiplied by the depth, so that pixels against the background (depth 1.0 I imagine) are blurred to the maximum allowed while closer pixels are blurred proportionally less. 

Which one is the intended look of this feature? The blurry look against the skysphere or the sharper look below?

**Comment by emanuele3d** [comment]
Also, assuming it is some kind of force field, shouldn't the back side of the sphere also visible?

**Comment by immortius** [comment]
@emanuele3d 
Not my concern whether it is double sided or not, or what is intended in that space. Although it does raise the question of whether double sided rendering is available for standard mesh - probably not because mesh are expected to have normals and normals don't work for double sided rendering.

As for how they should be blurred... probably easiest not to blur translucent pixels - so draw them after blurring. The "correct" handling is unlikely to be tenable.

**Comment by Cervator** [comment]
Thanks - talked to @skaldarnar who is using it, so merging it in :-)

**Comment by emanuele3d** [comment]
I understand that perhaps I should have explicitly flagged my comment as not really concerning this particular PR but only the issues the screenshot displays. 

I do not know the implementation details regarding rendering meshes, but by my understanding, in the context of rendering, normals tend to be useful (only?) for lighting purposes. So a shader that is not affected by lighting, such as the light-only effect displayed here for example, wouldn't be concerned by normals going the wrong way. But even fragment shaders that are lighting-conscious can use the gl_FrontFacing input (available since GLSL 1.10) to handle a normal going the wrong way, i.e. by reversing it if gl_FrontFacing is false. So, while the current implementation might place hurdles in using double-sided meshes, they might not be insurmountable.

Concerning the blur, never say never! ;)

**Comment by emanuele3d** [comment]
I thought about the blur a little more. I didn't realize the blur on the force-field thingy is actually the DoF. I thought the blur was part of the effect and the sharp portions of it were the non-intended part. 

So, in this case the whole thing should be sharp actually, but the DoF effect has no depth information from the translucent object, hence the blurring using background depth values or depth values from nearby blocks.

Hmmm. Again, not the concern of this PR, but perhaps:
* the shader might output depth values into a depth buffer (perhaps this is a opengl > 2.0 feature - gotta check) at least where the opacity is full, and perhaps even where the opacity is partial, i.e. up to a threshold.
    * alternatively the object might provide depth information as a fully opaque object - it'd be something at least, for non-doublesided objects.
* translucent objects would need the DoF applied to them *before* they are combined with the rest of the scene, which should also have had the DoF applied.

Perhaps this would generate artifacts in pixels that are semi-opaque. Or on the point of contact between translucent and opaque objects. But perhaps they can be minimized to the point of not being bothersome?

**Comment by Cervator** [comment]
All that is way beyond me but I realize DOF and blur in general still needs work. @IWhoI did the last round of tweaks to get us from one rough state to a slightly less rough state (less twitchy blur) but yeah we need more magic applied here.

If you have some constructive details and thoughts like the above you'd like to keep (I know I'd like to keep them, even if I don't understand them!) feel free to put in a new issue for future work, I fear we will forget details from here :-)

**Comment by IWhoI** [comment]
Hey guys. I hate not having the time to work on Terasology but I can attempt to help with the little I know. The problem doesn't seem to have anything to do with the translucency but that the translucent object isn't being hit by the trace used to calculate the focus distance and it's depth information is not in the depth buffer maybe? The trace is going right through it to the max distance and is blurring the scene based on the landscape distances in the depth buffer I think.

**Comment by immortius** [comment]
Ok, the thing to understand about the translucent mesh is they do not write to the depth buffer, but are tested against the depth buffer. The testing ensures the bits behind the opaque mesh are not rendered. The not-writing-to-depth-buffer is because they can overlap, and they are not ordered, so we cannot be sure that everything behind one translucent mesh is drawn before it is draw. It is a little more complicated than that since mesh can be in complex shapes such that parts of them are in-front and behind other mesh, and also there is no guarantee of the order the triangles within a mesh are rendered. I can imagine a complex system where mesh are automatically split apart and the parts ordered, but seems pretty complex and unlikely to perform well.

Also, even if we solved all this, it doesn't make sense that an extremely transparent mesh towards the user would cause the pixels behind it to inherit its depth.

So solutions...

If we rendered translucent mesh to a separate buffer recording depth but not depth testing, applied DoF separately to both the main buffer and the translucent buffer, and then combined the two, I think we would get a close enough to correct result? Perhaps also some tweeking around when a pixel is considered dominant enough to contribute to depth?

A lot of this overlaps with what @emanuele3d said.

**Comment by emanuele3d** [comment]
I thought about it some more. The most that we can do, I suspect is to:
1. render the scene without translucent objects and apply DOF.
2. render the front-facing triangles of translucent objects with everything else rendered in black. Apply DOF to the result by considering the translucent triangles opaque from a depth-buffer perspective.
3. repeat step 2 with back-facing polygons
4. combine the three separate passes 

However, as I thought about this, I also thought of situations in which it will break down. And fundamentally it boils down to this: when two or more triangles visually overlap in one of the two translucent passes, the pixels of the triangle closest to the camera will be DOF-blurred correctly, but the pixels of all the triangles behind the first will be blurred using the same depth-value of that first triangle. This would mean that a far away translucent object behind a nearby one, would be blurred as if it is nearby, and only in those portions of it that are behind the nearby object.

So, the specific case of the picture shown in this Issue could be addressed, as it is a simple convex surface. The general case of more complex surfaces with multiple layers of overlapping front-facing triangles or even just multiple translucent objects one behind another would generate problems similar to the image seen in the opening of this issue.

In short, Immortius had it right all along.

Unless.... we raytrace. AND... we simulate the way the rays move through the lenses of a real camera, the latter part being the scariest of the two, raytracing being pretty scary already.
